executor/src/witgen/expression_evaluator.rs

```diff
@@ -56,11 +56,21 @@ where
         op: &BinaryOperator,
         right: &'a Expression<T>,
     ) -> AffineResult<&'a PolynomialReference, T> {
-        match (self.evaluate(left), op, self.evaluate(right)) {
-            // Special case for multiplication: It is enough for one to be known zero.
-            (Ok(zero), BinaryOperator::Mul, _) | (_, BinaryOperator::Mul, Ok(zero))
-                if zero.constant_value() == Some(0.into()) =>
-            {
+        let left = self.evaluate(left);
+
+        // Short-circuit multiplication by zero.
+        if *op == BinaryOperator::Mul {
+            if let Ok(zero) = &left {
+                if zero.constant_value() == Some(0.into()) {
```

Comment:
Maybe this could be [is_zero](https://docs.rs/num/latest/num/traits/trait.Zero.html#tymethod.is_zero) and the comparison can be against a static `ZERO` instead of the dynamic `0.into()`

Reply:
Changed it to use is_zero